repl/src/main/scala/ammonite/repl/interp/Evaluator.scala

```diff
@@ -206,6 +222,74 @@ object Evaluator{
       )
     }
 
+    def compileCacheBlock(wrapperName: String, code: String, scriptImports: Seq[ImportData]): Res[(Class[_], Seq[ImportData])] = for {
+      _ <- Catching{ case e: ThreadDeath => interrupted() }
+      (classFiles, importData) <- compileClass(
+        s"""
+        $defaultImportBlock
+        ${importBlock(scriptImports )}
+
+        case object $wrapperName{
+
+          $code
+
+          def $$main() = {}
+        }
+        """
+      )
+      
+      _ = cacheSave(wrapperName,(classFiles,importData))
+      cls <- loadClass(wrapperName,classFiles) 
+    } yield (cls, importData)
+
+    def processScriptBlock(code: String, scriptImports: Seq[ImportData]) = {
+      val wrapperName = cacheTag(code, scriptImports)
+      loadCachedClass(wrapperName) match {
+        case Some((cls,importData)) => {
+          evalMain(cls)
+          Res.Success(
+            Evaluated(
+              wrapperName,
+              importData.map(id => id.copy(
+                wrapperName = wrapperName,
+                prefix = if (id.prefix == "") wrapperName else id.prefix
+              ))
+            )
+          )
+        }
+        case None => {
+          for {
+            (cls, newImports) <- compileCacheBlock(wrapperName, code, scriptImports)
+          } yield {
+            evalMain(cls)
+            Evaluated(
+              wrapperName,
+              newImports.map(id => id.copy(
+                wrapperName = wrapperName,
+                prefix = if (id.prefix == "") wrapperName else id.prefix
+              ))
+            )
+          }
+        }
+      }
+    }
+
+    def cacheTag(code: String, imports: Seq[ImportData]): String = {
```

Comment:
Good point, I just didn't notice it. Perhaps we could move `cacheTag` into the companion object and pass this stuff in as arguments in manually? I feel like that would make it clearer what the tagging does and does not depend on, since currently potentially it takes the whole evaluator state as input